dns-secondary-operator/lib/charms/dns_transfer/v0/dns_transfer.py

```diff
@@ -0,0 +1,392 @@
+            relation_data = relation.data[app]
+            for key, value in relation_data.items():
+                loaded_data[key] = json.loads(value)
+            return DNSTransferProviderData.model_validate(loaded_data)
+        except json.JSONDecodeError as ex:
+            raise ValueError from ex
+
+
+class DNSTransferRequirerData(pydantic.BaseModel):
+    """List of information for the requirer to manage and transfer.
+
+    Attributes:
+        addresses: IP list of the units composing the provider's deployment.
+    """
+
+    addresses: list[pydantic.IPvAnyAddress]
+
+    # pydantic wants 'self' as first argument
+    @pydantic.field_validator("addresses", mode="before")
+    def deduplicate_addresses(cls, v: Any) -> list:  # noqa: N805 pylint: disable=E0213
+        """Deduplicate addresses.
+
+        Args:
+            v: value
+
+        Returns:
+            list with unique values.
+        """
+        return list(set(v))
+
+    def to_relation_data(self) -> dict[str, str]:
+        """Convert an instance of DNSTransferRequirerData to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        return {"addresses": json.dumps(self.addresses, default=str)}
+
+    @classmethod
+    def from_relation_data(cls, relation: ops.Relation) -> "DNSTransferRequirerData":
+        """Initialize a new instance of the DNSTransferRequirerData class from the relation.
+
+        Args:
+            relation: the relation.
+
+        Returns:
+            A DNSTransferRequirerData instance.
+
+        Raises:
+            ValueError: if the value is not parseable.
+        """
+        try:
+            loaded_data = {}
+            app = cast(ops.Application, relation.app)
+            relation_data = relation.data[app]
+            for key, value in relation_data.items():
+                loaded_data[key] = json.loads(value)
+            return DNSTransferRequirerData.model_validate(loaded_data)
+        except json.JSONDecodeError as ex:
+            raise ValueError from ex
+
+
+class DNSTransferRequires(ops.Object):
+    """Requirer side of the DNS transfer requires relation."""
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the requirer charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+    def get_remote_relation_data(self) -> DNSTransferProviderData | None:
+        """Retrieve the remote relation data.
+
+        Returns:
+            DNSTransferProviderData: the relation data.
+        """
+        relation = self.model.get_relation(self.relation_name)
+        if not relation:
+            return None
+        return DNSTransferProviderData.from_relation_data(relation) if relation else None
```

Comment:
```suggestion
        return DNSTransferProviderData.from_relation_data(relation)
```
`relation` is already checked if it exists above.